changefeedccl: add timers around key parts of the changefeed pipeline

asg0451 opened the pull request:
Add timers around key parts of the changefeed
pipeline to help debug feeds experiencing issues.
We emit latency histograms for each stage using
the `changefeed.stage.<stage>.latency` metrics.
The metric respects the changefeed `scope` label
for debugging specific feeds.

The implementation exposes a mechanism to add new
timers for stages with ease.

Fixes: #121538

Release note (enterprise change): Added timers
around key parts of the changefeed pipeline to
help debug feeds experiencing issues. The
`changefeed.stage.<stage>.latency` metrics now
emit latency histograms for each stage. The metric
respects the changefeed `scope` label for
debugging specific feeds.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/128794)
<!-- Reviewable:end -->

asg0451 committed:
changefeedccl: add timers around key parts of the changefeed pipeline

Add timers around key parts of the changefeed
pipeline to help debug feeds experiencing issues.
We emit latency histograms for each stage using
the `changefeed.stage.<stage>.latency` metrics.
The metric respects the changefeed `scope` label
for debugging specific feeds.

The implementation exposes a mechanism to add new
timers for stages with ease.

Fixes: #121538

Release note (enterprise change): Added timers
around key parts of the changefeed pipeline to
help debug feeds experiencing issues. The
`changefeed.stage.<stage>.latency` metrics now
emit latency histograms for each stage. The metric
respects the changefeed `scope` label for
debugging specific feeds.

asg0451 commented:
@andyyang890 @wenyihu6 @rharding6373 do you guys have any suggestions for other important stages i should drop a timer in?

andyyang890 reviewed (commented):
Only taken a cursory look so far but I'll come back and do a more thorough review.

> Reviewed 9 of 19 files at r1, 2 of 5 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/128794#-:-O4MaMv_45BfQeVN-1aY:b-vb5wl8)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @aa-joshi, @angles-n-daemons, @arjunmahishi, @asg0451, and @wenyihu6)

___
*[`-- commits` line 11 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/128794#-O4MYO426pFiTQwMUfJT):*
nit: s/option/cluster setting/
___
*[`-- commits` line 27 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/128794#-O4MYSg71GsZobQhywx7):*
nit: add "cluster setting" to the last sentence
___
*[`pkg/ccl/changefeedccl/timers/timers.go` line 27 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/128794#-O4MaUkxAADhRUCsM-Ky:-O4MaUkxAADhRUCsM-Kz:b6zjha7) ([raw file](https://github.com/cockroachdb/cockroach/blob/4d9174895d67ea8f40f03936745f0280dac7bc15/pkg/ccl/changefeedccl/timers/timers.go#L27)):*
> ```Go
> // TODO: should we add a label for sink type? or other things maybe?
> 
> var TimersEnabled = settings.RegisterBoolSetting(
> ```

Do you think there'd be a really big performance impact with leaving the timers always enabled?
___
*[`pkg/ccl/changefeedccl/timers/timers.go` line 69 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/128794#-O4MgGQb39qVcgHtaNP5:-O4MgGQb39qVcgHtaNP6:b-ycw7o3) ([raw file](https://github.com/cockroachdb/cockroach/blob/4d9174895d67ea8f40f03936745f0280dac7bc15/pkg/ccl/changefeedccl/timers/timers.go#L69)):*
> ```Go
> 	Latencies  map[Stage]*aggmetric.AggHistogram
> 	settings   *settings.Values
> 	components struct {
> ```

nit: I think what I've generally seen is structs like this are named `mu` and then the inner fields have more descriptive names. This helps reinforce that you need to lock it before you read/write from the inner fields.
___
*[`pkg/ccl/changefeedccl/timers/timers.go` line 80 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/128794#-O4Ma4v95mNJzRDFuFgT:-O4Ma4vAAxl16aPd0nHP:bko0wp2) ([raw file](https://github.com/cockroachdb/cockroach/blob/4d9174895d67ea8f40f03936745f0280dac7bc15/pkg/ccl/changefeedccl/timers/timers.go#L80)):*
> ```Go
> func New(histogramWindow time.Duration, settings *settings.Values) *Timers {
> 	metaFor := func(stage Stage) metric.Metadata {
> 		baseMeta := metric.Metadata{Name: "changefeed.stage.latency", Help: "Latency of changefeed stage", Unit: metric.Unit_NANOSECONDS, Measurement: "Latency"}
> ```

imo this should be ordered `changefeed.latency.stage`, wdyt?
___
*[`pkg/ccl/changefeedccl/timers/timers.go` line 121 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/128794#-O4Mi4fpDwuf52pKix9K:-O4Mi4fpDwuf52pKix9L:b22upnb) ([raw file](https://github.com/cockroachdb/cockroach/blob/4d9174895d67ea8f40f03936745f0280dac7bc15/pkg/ccl/changefeedccl/timers/timers.go#L121)):*
> ```Go
> 	}
> 	if _, ok := st.Latencies[stage]; !ok {
> 		log.Warningf(context.Background(), "tried to start timer for unknown stage %s", stage)
> ```

It'd be nice to pass in the actual context since it'll include the changefeed job ID in the log
___
*[`pkg/util/metric/registry.go` line 151 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/128794#-O4MiLgx6RDEVD8rz5mU:-O4MiLgx6RDEVD8rz5mV:b-lhxlif) ([raw file](https://github.com/cockroachdb/cockroach/blob/4d9174895d67ea8f40f03936745f0280dac7bc15/pkg/util/metric/registry.go#L151)):*
> ```Go
> 		case reflect.Map:
> 			for mr := vfield.MapRange(); mr.Next(); {
> 				kelem, velem := mr.Key(), mr.Value()
> ```

nit: s/kelem/kElem/ and s/velem/vElem/


<!-- Sent from Reviewable.io -->

asg0451 reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/128794)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @aa-joshi, @andyyang890, @angles-n-daemons, @arjunmahishi, and @wenyihu6)

___
*[`pkg/ccl/changefeedccl/timers/timers.go` line 27 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/128794#-O4MaUkxAADhRUCsM-Ky:-O7Ag0uYD2CjhNm3G8Y6:b-inew5r) ([raw file](https://github.com/cockroachdb/cockroach/blob/4d9174895d67ea8f40f03936745f0280dac7bc15/pkg/ccl/changefeedccl/timers/timers.go#L27)):*
<details><summary><i>Previously, andyyang890 (Andy Yang) wrote…</i></summary><blockquote>

Do you think there'd be a really big performance impact with leaving the timers always enabled?
</blockquote></details>

Probably not, but I'm inclined to put new features behind a flag pretty much always in case they cause problems. We can look at setting this to default to on after it's proven itself stable and useful.
___
*[`pkg/ccl/changefeedccl/timers/timers.go` line 80 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/128794#-O4Ma4v95mNJzRDFuFgT:-O7AgRqu8hqqtA4E8HIH:b-6638yk) ([raw file](https://github.com/cockroachdb/cockroach/blob/4d9174895d67ea8f40f03936745f0280dac7bc15/pkg/ccl/changefeedccl/timers/timers.go#L80)):*
<details><summary><i>Previously, andyyang890 (Andy Yang) wrote…</i></summary><blockquote>

imo this should be ordered `changefeed.latency.stage`, wdyt?
</blockquote></details>

i think it should be ordered this way which is why i wrote it this way lol

do you have a good reason to change it? i think this is a common naming paradigm -- eg `sql.txn.latency`, `storage.wal.fsync.latency`, to pick 2 random ones
___
*[`pkg/ccl/changefeedccl/timers/timers.go` line 121 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/128794#-O4Mi4fpDwuf52pKix9K:-O7Ahkzw0nB-r0OCbsSg:b-nqhy4e) ([raw file](https://github.com/cockroachdb/cockroach/blob/4d9174895d67ea8f40f03936745f0280dac7bc15/pkg/ccl/changefeedccl/timers/timers.go#L121)):*
<details><summary><i>Previously, andyyang890 (Andy Yang) wrote…</i></summary><blockquote>

It'd be nice to pass in the actual context since it'll include the changefeed job ID in the log
</blockquote></details>

yeah true that's just annoying to do.. let me see if i can find a nice way to do it or an alternative
___
*[`pkg/util/metric/registry.go` line 151 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/128794#-O4MiLgx6RDEVD8rz5mU:-O7AhWDB5tMJoEHy1RES:bhqfuwt) ([raw file](https://github.com/cockroachdb/cockroach/blob/4d9174895d67ea8f40f03936745f0280dac7bc15/pkg/util/metric/registry.go#L151)):*
<details><summary><i>Previously, andyyang890 (Andy Yang) wrote…</i></summary><blockquote>

nit: s/kelem/kElem/ and s/velem/vElem/
</blockquote></details>

being consistent with the surrounding context trumps the default of camelCase -- see the rest of this function


<!-- Sent from Reviewable.io -->

andyyang890 reviewed (approved):
I like the simplifications! Needs a `./dev gen` to clean up the generated files and still an open question about ordering of the metric name parts but otherwise <img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 

> Reviewed 1 of 5 files at r2, 15 of 15 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/128794#-:-O7XNYen9YgF_buNZt5K:bnftqq)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @aa-joshi, @angles-n-daemons, @arjunmahishi, @asg0451, and @wenyihu6)

___
*[`pkg/ccl/changefeedccl/kvfeed/kv_feed.go` line 728 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/128794#-O7XQ7Zn5Ws9dZVZgJi_:-O7XQ7Zn5Ws9dZVZgJia:b-ru47m0) ([raw file](https://github.com/cockroachdb/cockroach/blob/62c7646e9fce1b8417168b7416c34386c7366914/pkg/ccl/changefeedccl/kvfeed/kv_feed.go#L728)):*
> does this deserve one?

imo probably not
___
*[`pkg/ccl/changefeedccl/kvfeed/physical_kv_feed.go` line 125 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/128794#-O7XQwUxAoP6giNpZznT:-O7XQwUxAoP6giNpZznU:b-b6ezq) ([raw file](https://github.com/cockroachdb/cockroach/blob/62c7646e9fce1b8417168b7416c34386c7366914/pkg/ccl/changefeedccl/kvfeed/physical_kv_feed.go#L125)):*
> or i could put this in the memBuf.Add method

I think it's good here
___
*[`pkg/ccl/changefeedccl/timers/timers.go` line 80 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/128794#-O4Ma4v95mNJzRDFuFgT:-O7XIdsTE3OFzh1exzbZ:b-aemggh) ([raw file](https://github.com/cockroachdb/cockroach/blob/4d9174895d67ea8f40f03936745f0280dac7bc15/pkg/ccl/changefeedccl/timers/timers.go#L80)):*
<details><summary><i>Previously, asg0451 (Miles Frankel) wrote…</i></summary><blockquote>

i think it should be ordered this way which is why i wrote it this way lol

do you have a good reason to change it? i think this is a common naming paradigm -- eg `sql.txn.latency`, `storage.wal.fsync.latency`, to pick 2 random ones
</blockquote></details>

hmm i think at the time i wrote the comment i was thinking that it belongs to a group of metrics that tracks latency so that should be the first part of the name, and also that it would be nicer if the "stage" was right beside the actual stage name. i see that you've changed that now and imo it would be a bit more natural to have a common prefix with the specific stage name at the end. 


<!-- Sent from Reviewable.io -->

asg0451 commented on pkg/ccl/changefeedccl/kvfeed/physical_kv_feed.go:
```diff
@@ -133,11 +140,13 @@ func (p *rangefeed) addEventsToBuffer(ctx context.Context) error {
 				if p.knobs.ShouldSkipCheckpoint != nil && p.knobs.ShouldSkipCheckpoint(t) {
 					continue
 				}
+				stop := p.st.RangefeedBufferCheckpoint.Start()
```
TODO: null ptr error here. what's nil?

andyyang890 reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/128794)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @aa-joshi, @angles-n-daemons, @arjunmahishi, @asg0451, and @wenyihu6)

___
*[`pkg/ccl/changefeedccl/timers/timers.go` line 80 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/128794#-O4Ma4v95mNJzRDFuFgT:-O7dyTeiCO1cr8HqaE_h:b1u7hx4) ([raw file](https://github.com/cockroachdb/cockroach/blob/4d9174895d67ea8f40f03936745f0280dac7bc15/pkg/ccl/changefeedccl/timers/timers.go#L80)):*
<details><summary><i>Previously, andyyang890 (Andy Yang) wrote…</i></summary><blockquote>

hmm i think at the time i wrote the comment i was thinking that it belongs to a group of metrics that tracks latency so that should be the first part of the name, and also that it would be nicer if the "stage" was right beside the actual stage name. i see that you've changed that now and imo it would be a bit more natural to have a common prefix with the specific stage name at the end.
</blockquote></details>

I've come around on having `.latency` as the last item. I'm fine with the names as-is.
___
*[`pkg/ccl/changefeedccl/timers/timers.go` line 35 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/128794#-O7dzWNL-tNF3wuc5Vhk:-O7dzWNL-tNF3wuc5Vhl:b-zg3096) ([raw file](https://github.com/cockroachdb/cockroach/blob/67150014993fe90c0e20195f0b9ee5df4110f5a3/pkg/ccl/changefeedccl/timers/timers.go#L35)):*
> ```Go
> 		baseMeta := metric.Metadata{Unit: metric.Unit_NANOSECONDS, Measurement: "Latency"}
> 		baseMeta.Name = fmt.Sprintf("changefeed.stage.%s.latency", stage)
> 		baseMeta.Help = fmt.Sprintf("Latency of the changefeed stage: %s", stage)
> ```

minor nit: should the help text version of the stage name replace the underscores with spaces maybe?


<!-- Sent from Reviewable.io -->

asg0451 reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/128794)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @aa-joshi, @andyyang890, @angles-n-daemons, @arjunmahishi, and @wenyihu6)

___
*[`pkg/ccl/changefeedccl/timers/timers.go` line 35 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/128794#-O7dzWNL-tNF3wuc5Vhk:-O7e9-9QBu9nYCDH85K3:b-by1653) ([raw file](https://github.com/cockroachdb/cockroach/blob/67150014993fe90c0e20195f0b9ee5df4110f5a3/pkg/ccl/changefeedccl/timers/timers.go#L35)):*
<details><summary><i>Previously, andyyang890 (Andy Yang) wrote…</i></summary><blockquote>

minor nit: should the help text version of the stage name replace the underscores with spaces maybe?
</blockquote></details>

i'm not sure that would really make things clearer. i could write descriptive names with capitalization and more words but what's the roi on that?


<!-- Sent from Reviewable.io -->

asg0451 reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/128794)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @aa-joshi, @andyyang890, @angles-n-daemons, @arjunmahishi, and @wenyihu6)

___
*[`pkg/ccl/changefeedccl/timers/timers.go` line 35 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/128794#-O7dzWNL-tNF3wuc5Vhk:-O7eBF65AP712N8RUwqn:b-gvogjx) ([raw file](https://github.com/cockroachdb/cockroach/blob/67150014993fe90c0e20195f0b9ee5df4110f5a3/pkg/ccl/changefeedccl/timers/timers.go#L35)):*
<details><summary><i>Previously, asg0451 (Miles Frankel) wrote…</i></summary><blockquote>

i'm not sure that would really make things clearer. i could write descriptive names with capitalization and more words but what's the roi on that?
</blockquote></details>

update: i did that


<!-- Sent from Reviewable.io -->

asg0451 commented:
bors r=andyyang890
